src/main/scala/coupledL2/tl2chi/MSHRCtl.scala

```diff
@@ -222,5 +223,30 @@ class MSHRCtl(implicit p: Parameters) extends TL2CHIL2Module with HasCHIOpcodes
       XSPerfMax("mshr_latency", timer, enable)
     }
   }*/
+
+  val hpm_timers = RegInit(VecInit(Seq.fill(mshrsAll)(0.U(10.W))))
+  // TODO: Is the width(4.W) proper here?
+  val lmiss = Wire(Vec(mshrsAll, UInt(4.W)))
+  for (((hpm_timer, m), i) <- hpm_timers.zip(mshrs).zipWithIndex) {
+    when (m.io.alloc.valid) {
+      hpm_timer := 1.U
+    }.otherwise {
+      hpm_timer := hpm_timer + 1.U
+    }
+    val enable = m.io.status.valid && m.io.status.bits.will_free
+    when (enable && hpm_timer > 200.U) {
+      lmiss(i) := 1.U
+    }.otherwise {
+      lmiss(i) := 0.U
+    }
+  }
+
+  val perfEvents = Seq(
+    ("l2_cache_refill", io.resps.rxdat.valid && (io.resps.rxdat.respInfo.chiOpcode.get === CompData) && io.resps.rxdat.respInfo.last),
```

Comment:
Refill should also consider separate data and response. And what about dateless request?

Reply:
Fixed in https://github.com/OpenXiangShan/CoupledL2/pull/251/commits/1d073e51d3c51aff2609ce1c6dbb9705bf6cb063. Dataless requests aren't included here.